===========================================Tiền Xử Lí Và Khám Phá Dữ Liệu=================================

1. IMPORT THƯ VIỆN

In [1]:

import numpy as np
import pandas as pd

2. ĐỌC DỮ LIỆU KIỂM TRA VÀ KIỂM TRA SỐ DÒNG, SỐ CỘT DỮ LIỆU

In [3]:
df = pd.read_csv("../Data/data.csv")
df.head()

Country  Year Fossil fuels         Gas        Coal         Oil  Solar  \
0     Algeria  2000      300 TWh  191.01 TWh    5.64 TWh  103.13 TWh  0 TWh   
1   Argentina  2000      595 TWh  322.97 TWh    7.51 TWh  264.94 TWh  0 TWh   
2   Australia  2000    1,253 TWh  205.55 TWh  591.66 TWh  455.98 TWh  0 TWh   
3     Austria  2000      262 TWh   78.21 TWh   42.47 TWh  141.75 TWh  0 TWh   
4  Azerbaijan  2000      128 TWh   52.83 TWh    0.00 TWh   74.69 TWh  0 TWh   

    Wind Renewables Hydropower  
0  0 TWh      0 TWh      0 TWh  
1  0 TWh     86 TWh     84 TWh  
2  0 TWh     52 TWh     49 TWh  
3  0 TWh    129 TWh    124 TWh  
4  0 TWh      5 TWh      5 TWh

In [4]:
shape = df.shape
print(f"Shape: {shape}")

Shape: (1840, 10)


In [6]:
#Số dòng
if shape[0] > 1000:
    print(f"Số dòng dữ liệu đạt yêu cầu.")
else:
    print(f"Số dòng dữ liệu không đạt yêu cầu")
#Số cột
if shape[1] >= 5:
    print(f"Số cột dữ liệu đạt yêu cầu.")
else:
    print(f"Số cột dữ liệu không đạt yêu cầu")
#Kiểm tra có dòng duplicate không
num_duplicated_rows = df.index.duplicated().sum()
if num_duplicated_rows == 0:
    print(f"Dữ liệu không có dòng lặp lại.")
else:
    print(f"Dữ liệu bạn có {num_duplicated_rows} lặp lại.")

Số dòng dữ liệu đạt yêu cầu.
Số cột dữ liệu đạt yêu cầu.
Dữ liệu không có dòng lặp lại.


3. Ý NGHĨA CỦA CÁC CỘT TRONG DATAFRAME VÀ THÔNG TIN VỀ DATAFRAME

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1840 entries, 0 to 1839
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Country       1840 non-null   object
 1   Year          1840 non-null   int64 
 2   Fossil fuels  1697 non-null   object
 3   Gas           1752 non-null   object
 4   Coal          1762 non-null   object
 5   Oil           1817 non-null   object
 6   Solar         1805 non-null   object
 7   Wind          1817 non-null   object
 8   Renewables    1639 non-null   object
 9   Hydropower    1639 non-null   object
dtypes: int64(1), object(9)
memory usage: 143.9+ KB


- Ý nghĩa về các cột trong DataFrame:
    **Chú ý:** các đơn vị năng lượng sử dụng ở đây là TWh, vì đây là lượng năng lượng sử dụng cả một năm.
    + Country: Đất nước mà chúng ta đang đề cập đến.
    + Year: Năm tương ứng.
    + Fosil fuels: Tổng năng lượng hóa thạch sử dụng của đất nước Country trong năm Year (nó bao gồm Oil,Coal ở phía dưới).
    + Gas: Tổng năng lượng từ khí gas sử dụng của đất nước Country trong năm Year.
    + Coal: Tổng năng lượng từ than đá sử dụng của đất nước Country trong năm Year.
    + Oil: Tổng năng lượng từ dầu sử dụng của đất nước Country trong năm Year.
    + Solar: Tổng năng lượng mặt trời sử dụng của đất nước Country trong năm Year.
    + Wind: Tổng năng lượng gió sử dụng của đất nước Country trong năm Year.
    + Renewables: Tổng năng lượng tái tạo sử dụng của đất nước Country trong năm Year (nó bao gồm Wind,Solar,Hydropower).
    + Hydropower: Tổng năng lượng từ thủy điện sử dụng của đất nước Country trong năm Year.

- Tổng quan về dữ liệu:
    + Dữ liệu có 1840 dòng và 9 cột.
    + Có 1 đối tượng kiểu in64 là Year, còn lại là kiểu object
    + Số lượng dữ liệu không phải null thể hiện như ở trên.

4. XỬ LÝ DỮ LIỆU, CHUYỂN CÁC KIỂU DỮ LIỆU CẦN THIẾT VỀ NUMERIC VÀ XEM XÉT BỎ CÁC CỘT CÓ ĐỘ MẤT MÁT DỮ LIỆU LỚN


        4.1 Xử lý dữ liệu, chuyển các kiểu dữ liệu về kiểu numeric

In [8]:
#Chuyển các dữ liệu về kiểu số.
convert_columns=df.columns[2:]#Bỏ ra cột tên nước và năm đi
for col in convert_columns:
    df[col] = pd.to_numeric(df[col].str.replace(' TWh', '').str.replace(',', ''),errors='coerce')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1840 entries, 0 to 1839
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Country       1840 non-null   object 
 1   Year          1840 non-null   int64  
 2   Fossil fuels  1697 non-null   float64
 3   Gas           1752 non-null   float64
 4   Coal          1762 non-null   float64
 5   Oil           1817 non-null   float64
 6   Solar         1805 non-null   float64
 7   Wind          1817 non-null   float64
 8   Renewables    1639 non-null   float64
 9   Hydropower    1639 non-null   float64
dtypes: float64(8), int64(1), object(1)
memory usage: 143.9+ KB


- Sau khi thực hiện như trên ta thấy các dữ liệu các dữ liệu năng lượng đều chuyển về kiểu float64.

        4.2 Tìm kiếm các cột dữ liệu bị mất quá nhiều.

In [10]:
#Tính toán giá trị phần trăm mà dữ liệu bị mất và hiển thị thêm thông tin về giá trị lớn nhất và nhỏ nhất luôn.
raw_df = df.select_dtypes(exclude=['object']) # chỉ lấy kiểu numeric, ko lấy kiểu object,str
def missing_ratio(raw_df):
    return raw_df.isna().mean() * 100
num_col_info_df = raw_df.agg([missing_ratio, "min", "mean", "max"]).round(1)
num_col_info_df

Year  Fossil fuels     Gas     Coal      Oil   Solar    Wind  \
missing_ratio     0.0           7.8     4.8      4.2      1.2     1.9     1.2   
min            2000.0          12.0     0.0      0.0      7.2     0.0     0.0   
mean           2011.0        1570.7   411.3    511.6    584.9     8.9    21.9   
max            2022.0       36319.0  8812.1  24559.5  11214.1  1115.0  1988.0   

               Renewables  Hydropower  
missing_ratio        10.9        10.9  
min                   0.0         0.0  
mean                184.7       123.4  
max                7093.0      3471.0

- Ở bảng trên ta thấy dòng đầu tiên là tỉ lệ phần trăm dữ liệu bị mất, tiếp theo là các dòng giá trị nhỏ nhất, trung bình và lớn nhất của tương ứng với từng cột thuộc tính.
- Cột year ở đây không có ý nghĩa nhưng mà ta thêm vào để thấy được dữ liệu bị mất của nó là 0% chứng tỏ chúng ta đang làm đúng.

        4.3 Xử lí nếu các cột dữ liệu bị mất quá nhiều